server/internal/infrastructure/gcs/file.go

```diff
@@ -0,0 +1,212 @@
+package gcs
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/url"
+	"path"
+	"strings"
+
+	"cloud.google.com/go/storage"
+	"github.com/kennygrant/sanitize"
+	"github.com/reearth/reearth-cms/server/internal/usecase/gateway"
+	"github.com/reearth/reearth-cms/server/pkg/file"
+	"github.com/reearth/reearth-cms/server/pkg/id"
+	"github.com/reearth/reearth-cms/server/pkg/log"
+	"github.com/reearth/reearth-cms/server/pkg/rerror"
+)
+
+const (
+	gcsAssetBasePath string = "assets"
+	fileSizeLimit    int64  = 1024 * 1024 * 100 // about 100MB
+)
+
+type fileRepo struct {
+	bucketName   string
+	base         *url.URL
+	cacheControl string
+}
+
+func NewFile(bucketName, base string, cacheControl string) (gateway.File, error) {
+	if bucketName == "" {
+		return nil, errors.New("bucket name is empty")
+	}
+
+	var u *url.URL
+	if base == "" {
+		base = fmt.Sprintf("https://storage.googleapis.com/%s", bucketName)
+	}
+
+	var err error
+	u, _ = url.Parse(base)
+	if err != nil {
+		return nil, errors.New("invalid base URL")
+	}
+
+	return &fileRepo{
+		bucketName:   bucketName,
+		base:         u,
+		cacheControl: cacheControl,
+	}, nil
+}
+
+func (f *fileRepo) ReadAsset(ctx context.Context, name string) (io.ReadCloser, error) {
+	sn := sanitize.Path(name)
+	if sn == "" {
+		return nil, rerror.ErrNotFound
+	}
+	return f.read(ctx, path.Join(gcsAssetBasePath, sn))
+}
+
+func (f *fileRepo) UploadAsset(ctx context.Context, file *file.File) (*url.URL, error) {
+	if file == nil {
+		return nil, gateway.ErrInvalidFile
+	}
+	if file.Size >= fileSizeLimit {
+		return nil, gateway.ErrFileTooLarge
+	}
+
+	sn := sanitize.Path(newAssetID() + path.Ext(file.Path))
+	if sn == "" {
+		return nil, gateway.ErrInvalidFile
+	}
+
+	filename := path.Join(gcsAssetBasePath, sn)
```

Comment:
specification is not followed

`gs://{bucket name}/{first 2 chars of uuid}/{rest of uuid}/{filename with ext}`